File: tools/clients/smoke.go
package main

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"runtime"
)

func oneMatch(pattern string) (string, error) {
	names, err := filepath.Glob(pattern)
	if err != nil {
		return "", err
	}
	if len(names) != 1 {
		return "", fmt.Errorf("expected one artifact matching %s, found %d", pattern, len(names))
	}
	return names[0], nil
}

func (t *tool) smoke(ctx context.Context) error {
	artifacts := t.path("dist/clients/" + t.config.Version)
	if err := verifyChecksums(artifacts); err != nil {
		return err
	}
	for _, name := range []string{"openapi.json", "openapi.zh-CN.json"} {
		same, err := samePath(filepath.Join(artifacts, name), t.path("api/"+name))
		if err != nil {
			return err
		}
		if !same {
			return fmt.Errorf("stale contract archive: %s", name)
		}
	}
	var operations object
	if err := readJSON(t.path("tools/clients/operations.json"), &operations); err != nil {
		return err
	}
	count := len(operations)
	work, err := os.MkdirTemp("", "wave-install-smoke-*")
	if err != nil {
		return err
	}
	defer os.RemoveAll(work)
	if err = t.run(ctx, work, nil, "uv", "venv", filepath.Join(work, "venv")); err != nil {
		return err
	}
	python := filepath.Join(work, "venv/bin/python")
	if runtime.GOOS == "windows" {
		python = filepath.Join(work, "venv/Scripts/python.exe")
	}
	wheel, err := oneMatch(filepath.Join(artifacts, "*.whl"))
	if err != nil {
		return err
	}
	if err = t.run(ctx, work, nil, "uv", "pip", "install", "--python", python, wheel); err != nil {
		return err
	}
	pythonCheck := fmt.Sprintf(`from wave_ai import Client, AsyncClient, OPERATIONS; from wave_ai_generated import AgentsApi; assert len(OPERATIONS)==%d; assert Client().raw().configuration.host == "http://localhost:8080"`, count)
	if err = t.run(ctx, work, nil, python, "-c", pythonCheck); err != nil {
		return err
	}
	if err = writeJSON(filepath.Join(work, "package.json"), object{"private": true, "type": "module"}); err != nil {
		return err
	}
	npm, err := oneMatch(filepath.Join(artifacts, "*.tgz"))
	if err != nil {
		return err
	}
	if err = t.run(ctx, work, nil, "npm", "install", "--ignore-scripts", npm); err != nil {
		return err
	}
	packageName, _ := json.Marshal(t.config.NPMPackage)
	js := fmt.Sprintf(`import {Client,operations,generated} from %s; if(Object.keys(operations).length!==%d || !new generated.AgentsApi(new Client().rawConfiguration())) process.exit(1);`, packageName, count)
	if err = t.run(ctx, work, nil, "node", "--input-type=module", "-e", js); err != nil {
		return err
	}
	name, extension := "wavectl", ".tar.gz"
	if runtime.GOOS == "windows" {
		name += ".exe"
		extension = ".zip"
	}
	archive := filepath.Join(artifacts, fmt.Sprintf("wavectl_%s_%s_%s%s", t.config.Version, runtime.GOOS, runtime.GOARCH, extension))
	data, err := archiveMember(archive, name)
	if err != nil {
		return err
	}
	binary := filepath.Join(work, name)
	if err = writeFile(binary, data, 0755); err != nil {
		return err
	}
	output, err := t.output(ctx, work, map[string]string{"WAVE_CONFIG_FILE": filepath.Join(work, "cli-config.json"), "WAVE_PROFILE": ""}, binary, "schema")
	if err != nil {
		return err
	}
	var schema object
	if err = json.Unmarshal([]byte(output), &schema); err != nil {
		return err
	}
	if len(obj(schema["operations"])) != count {
		return fmt.Errorf("CLI archive operation count differs from contract")
	}
	for _, skill := range []string{"wave-client", "wave-admin"} {
		archive := filepath.Join(artifacts, skill+"_"+t.config.Version+".zip")
		files, err := archiveFiles(t.path("skills/"+skill), "", nil)
		if err != nil {
			return err
		}
		for _, file := range files {
			want, err := os.ReadFile(file.source)
			if err != nil {
				return err
			}
			got, err := archiveMember(archive, file.name)
			if err != nil {
				return err
			}
			if !bytes.Equal(got, want) {
				return fmt.Errorf("stale skill archive: %s/%s", skill, file.name)
			}
		}
	}
	fmt.Fprintln(t.out, "Built CLI, Python wheel, npm package and Skill archives passed installation smoke checks")
	return nil
}

File: tools/clients/smoke.py
#!/usr/bin/env python3
"""Install the built Python/npm packages and execute the native CLI archive."""
import hashlib
import json
from pathlib import Path
import platform
import subprocess
import tarfile
import tempfile
import zipfile

ROOT=Path(__file__).resolve().parents[2]
version=json.loads((ROOT/'tools/clients/config.json').read_text())['version']
artifacts=ROOT/'dist/clients'/version
def run(*args,cwd=ROOT): subprocess.run(args,check=True,cwd=cwd)
for line in (artifacts/'SHA256SUMS').read_text().splitlines():
    checksum,name=line.split('  ',1)
    assert hashlib.sha256((artifacts/name).read_bytes()).hexdigest()==checksum,name
for name in ('openapi.json','openapi.zh-CN.json'):
    assert (artifacts/name).read_bytes()==(ROOT/'api'/name).read_bytes(),f'stale contract archive: {name}'
with tempfile.TemporaryDirectory(prefix='wave-install-smoke-') as directory:
    work=Path(directory)
    run('uv','venv',str(work/'venv'))
    python=work/'venv/bin/python'
    run('uv','pip','install','--python',str(python),str(next(artifacts.glob('*.whl'))))
    run(str(python),'-c','from wave_ai import Client, AsyncClient, OPERATIONS; from wave_ai_generated import AgentsApi; assert len(OPERATIONS)==54; assert Client().raw().configuration.host == "http://localhost:8080"',cwd=work)
    (work/'package.json').write_text('{"private":true,"type":"module"}')
    package=json.loads((ROOT/'sdks/typescript/package.json').read_text())['name']
    run('npm','install','--ignore-scripts',str(next(artifacts.glob('*.tgz'))),cwd=work)
    run('node','--input-type=module','-e',f'import {{Client,operations,generated}} from {json.dumps(package)}; if(Object.keys(operations).length!==54 || !new generated.AgentsApi(new Client().rawConfiguration())) process.exit(1);',cwd=work)
    system={'Linux':'linux','Darwin':'darwin','Windows':'windows'}[platform.system()]
    arch={'x86_64':'amd64','AMD64':'amd64','aarch64':'arm64','arm64':'arm64'}[platform.machine()]
    filename=f'wavectl_{version}_{system}_{arch}'
    if system=='windows':
        with zipfile.ZipFile(artifacts/(filename+'.zip')) as z: z.extract('wavectl.exe',work)
        binary=work/'wavectl.exe'
    else:
        binary=work/'wavectl'
        with tarfile.open(artifacts/(filename+'.tar.gz')) as tar: binary.write_bytes(tar.extractfile('wavectl').read())
        binary.chmod(0o755)
    output=subprocess.check_output([str(binary),'schema'],text=True)
    assert len(json.loads(output)['operations'])==54
    for skill in ('wave-client','wave-admin'):
        with zipfile.ZipFile(artifacts/f'{skill}_{version}.zip') as z:
            assert z.read('SKILL.md').startswith(f'---\nname: {skill}\n'.encode())
print('Built CLI, Python wheel, npm package and Skill archives passed installation smoke checks')
